fetcher.go

```diff
@@ -98,8 +114,43 @@ func (f *Fetcher) BlockMatching(ctx context.Context, root cid.Cid, match selecto
 	return results, errors
 }
 
-func (f *Fetcher) BlockAll(ctx context.Context, root cid.Cid) (chan FetchResult, chan error) {
-	ssb := builder.NewSelectorSpecBuilder(basicnode.Prototype__Any{})
+// Block fetches a schemaless node graph corresponding to single block by link.
+func Block(ctx context.Context, f Fetcher, link ipld.Link) (ipld.Node, error) {
+	prototype, err := prototypeFromLink(link)
+	if err != nil {
+		return nil, err
+	}
+	return f.BlockOfType(ctx, link, prototype)
+}
+
+// BlockMatching traverses a schemaless node graph starting with the given link using the given selector and possibly crossing
+// block boundaries. Each matched node is sent to the FetchResult channel.
+func BlockMatching(ctx context.Context, f Fetcher, root ipld.Link, match selector.Selector) (<-chan FetchResult, <-chan error) {
+	prototype, err := prototypeFromLink(root)
+	if err != nil {
+		errors := make(chan error, 1)
+		errors <- err
```

Comment:
should errors get closed after the error is sent here?
document who's responsible for closing the channels & semantics thereof

Reply:
I simplified the return types to a single channel with an error field in the result. Processing the two channels was already complicated, and I realized there's a potential race between an error event and closing the results that would require even more complexity on the consuming side to resolve. I've documented the (now simpler) channel behavior in the Fetcher interface.

Reply:
Actually, I changed it back to make it easier to process context cancelled events. This implementation should be solid though and it's now documented in the interface.